### ISSUE TITLE
train --recover doesn't work with GPU + Adam optimizer

### ISSUE THREAD

**Issue by joelgrus** [allenai/allennlp#998]
crashes when it tries to resume training

```
2018-03-19 13:41:54,514 - INFO - allennlp.data.dataset_readers.reading_comprehension.triviaqa - sample this iteration True
Traceback (most recent call last):
  File "/home/joelg/miniconda3/envs/allennlp2/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/joelg/miniconda3/envs/allennlp2/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/joelg/allennlp/allennlp/run.py", line 18, in <module>
    main(prog="python -m allennlp.run")
  File "/home/joelg/allennlp/allennlp/commands/__init__.py", line 73, in main
    args.func(args)
  File "/home/joelg/allennlp/allennlp/commands/train.py", line 100, in train_model_from_args
    args.recover)
  File "/home/joelg/allennlp/allennlp/commands/train.py", line 130, in train_model_from_file
    return train_model(params, serialization_dir, file_friendly_logging, recover)
  File "/home/joelg/allennlp/allennlp/commands/train.py", line 303, in train_model
    metrics = trainer.train()
  File "/home/joelg/allennlp/allennlp/training/trainer.py", line 649, in train
    train_metrics = self._train_epoch(epoch)
  File "/home/joelg/allennlp/allennlp/training/trainer.py", line 458, in _train_epoch
    self._optimizer.step()
  File "/home/joelg/miniconda3/envs/allennlp2/lib/python3.6/site-packages/torch/optim/adam.py", line 69, in step
    exp_avg.mul_(beta1).add_(1 - beta1, grad)
TypeError: add_ received an invalid combination of arguments - got (float, torch.cuda.FloatTensor), but expected one of:
 * (float value)
 * (torch.FloatTensor other)
 * (torch.SparseFloatTensor other)
 * (float value, torch.FloatTensor other)
      didn't match because some of the arguments have invalid types: (float, torch.cuda.FloatTensor)
 * (float value, torch.SparseFloatTensor other)
      didn't match because some of the arguments have invalid types: (float, torch.cuda.FloatTensor)
```

**Comment by matt-gardner** [comment]
I'm not certain what the issue is, but I'm guessing it's because of this: https://github.com/allenai/allennlp/blob/0c532d50ac52bca48bf9906864daf6735d5d02de/allennlp/training/trainer.py#L843-L850

@matt-peters, @DeNeutoy, any idea what's up?

**Comment by DeNeutoy** [comment]
Ah, this is because `load_state_dict` has different semantics for optimizers and modules. Optimizers are just doing a deepcopy of the state, into the optimizer, whereas the Module `load_state_dict` does an in-place copy into the parameters individually, which includes an implicit GPU transfer if the parameters are on different devices, I think.

**Comment by matt-peters** [comment]
What Mark said.  Here is a relevant thread:  https://github.com/pytorch/pytorch/issues/2830
Probably the easiest fix is to remove the device mapping when loading the optimizer.  This may still cause large models to fill up GPU memory due to all the extra copies in memory but I don't see another obvious work around.

**Comment by matt-peters** [comment]
Actually it looks like this is fixed on pytorch master -- https://github.com/pytorch/pytorch/blob/master/torch/optim/optimizer.py#L111

**Comment by matt-peters** [comment]
Fixed in #1144
